### PR TITLE
Fix Issue 21586 - Struct dtor is called twice if struct is created in…

### PR DESCRIPTION
**Pull request by BorisCarvajal**
…side ternary operator

Fix: Rewrite the CondExp in any case.

### CODE REVIEW

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
This should target stable.

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -8757,22 +8757,28 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                    // https://issues.dlang.org/show_bug.cgi?id=21586
+                    // Rewrite CondExp or e1 will miss direct construction, e.g.
+                    // e1 = a ? S(1) : ...; -> AST: e1 = a ? (S(0)).this(1) : ...;
+                    // a temporary created and an extra destructor call.
+                    // AST will be rewritten to:
+                    // a ? e1 = 0, e1.this(1) : ...; -> blitting plus construction
+                    if (e2x.op == TOK.question)
+                    {
+                        /* Rewrite as:
+                         *  a ? e1 = b : e1 = c;
```
Comment (kinke) [inline src/dmd/expressionsem.d:8769]: Does this also change the evaluation order? `myDest() = myCond() ? a : b;` -> `myDest` must still be called before `myCond`.

Comment (BorisCarvajal) [reply src/dmd/expressionsem.d:8769]: But isn't that an `AssignExp`? Here we are inside `ConstructExp` logic.

Comment (kinke) [reply src/dmd/expressionsem.d:8769]: Right; I guess it's unlikely `e1` has side effects, but it probably cannot be ruled out.

**Comment by pbackus** [comment]
This PR introduced a regression.
https://issues.dlang.org/show_bug.cgi?id=22226

**Comment by kinke** [comment]
It's not the culprit though (works with LDC), it's just uncovered a DMD backend bug apparently.

**Comment by BorisCarvajal** [comment]
Fix #12995

**Pull request merged**